src/main/scala/com/fulcrumgenomics/commons/async/Async.scala

```diff
@@ -0,0 +1,397 @@
+      * or [[RuntimeException]] as appropriate.
+      */
+    protected final def checkAndRethrow(): Unit = {
+      this.throwable.getAndSet(None).foreach { thr =>
+        this.isDone.set(true) // Ensure no further attempts to run
+        thr match {
+          case e: Error             => throw e
+          case ex: RuntimeException => throw ex
+          case t                    => throw new RuntimeException(t)
+        }
+      }
+    }
+  }
+
+
+  /** An asynchronous wrapper for an [[Iterator]].  A separate thread will be created to consume the source iterator.
+    * Will buffer up to [[bufferSize]] elements before the blocking when consuming the source.
+    *
+    * @param source the source iterator to consume
+    * @param bufferSize the maximum number of elements to buffer before the blocking when consuming the source, or None
+    *                   if unbounded.
+    * @tparam T the type of object to consume
+    */
+  class AsyncIterator[T](val source: Iterator[T], bufferSize: Option[Int] = None) extends Iterator[T] with AsyncThread {
+    bufferSize.foreach(size => require(size > 0, s"bufferSize must be greater than zero when given, found $size"))
+
+    private val queue: BlockingQueue[T] = bufferSize match {
+      case Some(size) => new ArrayBlockingQueue[T](size)
+      case None       => new LinkedBlockingDeque[T]()
+    }
+    private val thread = buildThread(getClass.getSimpleName)
+    private val outer  = this
+
+    // Start the thread after creating the queue
+    this.thread.start()
+
+    protected def run(): Unit = outer.source.foreach(outer.queue.put)
+
+    override protected def uponFinally(): Unit = outer.isDone.set(true)
+
+    /** Returns true if there exists more elements, false otherwise */
+    def hasNext(): Boolean = {
+      checkAndRethrow()
+
+      // If the queue is empty but the underlying thread is not done, then there may or may not be new
+      // items that will be put in the qeue.
+      while (this.queue.isEmpty && !this.isDone.get()) {
+        Thread.sleep(50)
+        checkAndRethrow() // check if there was any exception in the underlying thread
+      }
+
+      // At this point, either there are items in the queue, or the underlying thread has consumed the source
+      if (!this.queue.isEmpty) true
+      else { // the queue is empty, and the source has been consumed
+        require(this.isDone.get(), "Bug: iterator was not done when the thread was joined.")
+        tryAndModifyInterruptedException("Interrupted waiting on iterator thread to join.") { this.thread.join() }
+        checkAndRethrow()
+        !this.queue.isEmpty
+      }
+    }
+
+    /** Gets the next item. */
+    def next(): T = {
+      checkAndRethrow()
+      if (this.hasNext()) {
+        val item = tryAndModifyInterruptedException("Interrupted waiting on taking from the queue.") { this.queue.take() }
+        checkAndRethrow()
```

Comment:
I'm not sure you need to call this multiple times per method.  I think once at the start will do.